From 54dc36c450a09d5f41c0bc31b59247063ed30ed2 Mon Sep 17 00:00:00 2001
From: Alexey Spiridonov <lesha@fb.com>
Date: Wed, 4 Nov 2020 08:43:54 -0800
Subject: [PATCH] Fuse OSS library `protocol` into `thriftprotocol`

Summary:
I am trying to have Bistro use `find_package(FBThrift CONFIG REQUIRED)` to depend on FBThrift.

Unfortunately, if I do that, and link against `FBThrift::thrift2`, my build fails to link with:

```
.../libprotocol.so.1.0.0: undefined reference to `apache::thrift::CompactProtocolReader::throwBadType(unsigned char)'
.../libprotocol.so.1.0.0: undefined reference to `apache::thrift::CompactProtocolReader::readFieldBeginWithStateMediumSlow(apache::thrift::CompactProtocolReader::StructReadState&, short)'
collect2: error: ld returned 1 exit status
```

This is due to a cyclic dependency between `protocol` (aka Thrift1) and `thriftprotocol` (aka Thrift2). Specifically:
 - `thriftprotocol` explicitly depends on `protocol`
 - `protocol` has an undeclared dep on `thriftprotocol` because it includes the Thrift2 build of the "reflection" interface.

IMO, the "right" fix would be to make `STATIC` all "internal" Thrift libraries, and export only a `libthrift2.so` for external consumption.

However, I don't have cycles to update the entire Thrift build to do this.

Therefore, fusing the circularly dependent libraries is the path of least resistance to letting my project depend on Thrift.

Reviewed By: nlaptev, yfeldblum

Differential Revision: D24685618

fbshipit-source-id: a679d7b3556d1c9859b64f69f98bfa6d48398b1d
---
 thrift/lib/cpp/CMakeLists.txt  | 27 ++-------------------------
 thrift/lib/cpp2/CMakeLists.txt | 13 +++++++++++++
 2 files changed, 15 insertions(+), 25 deletions(-)

diff --git a/thrift/lib/cpp/CMakeLists.txt b/thrift/lib/cpp/CMakeLists.txt
index 763e9b844..516fff445 100644
--- a/thrift/lib/cpp/CMakeLists.txt
+++ b/thrift/lib/cpp/CMakeLists.txt
@@ -12,7 +12,6 @@
 # See the License for the specific language governing permissions and
 # limitations under the License.
 
-get_property(reflection-cpp2-SOURCES GLOBAL PROPERTY reflection-cpp2-SOURCES)
 
 add_library(
   thrift-core
@@ -50,27 +49,6 @@ target_link_libraries(
     ${GLOG_LIBRARIES}
 )
 
-bypass_source_check("${reflection-cpp2-SOURCES}")
-add_library(
-  protocol
-
-  protocol/TDebugProtocol.cpp
-  protocol/TJSONProtocol.cpp
-  protocol/TBase64Utils.cpp
-  protocol/TProtocolException.cpp
-  protocol/TSimpleJSONProtocol.cpp
-  ${reflection-cpp2-SOURCES}
-)
-add_dependencies(protocol reflection-cpp2-target)
-target_link_libraries(
-  protocol
-  PUBLIC
-    fmt::fmt
-    Folly::folly
-    ${GLOG_LIBRARIES}
-    thrift-core
-)
-
 add_library(
   transport
 
@@ -135,7 +113,7 @@ target_link_libraries(
   INTERFACE
     async
     concurrency
-    protocol
+    thriftprotocol
     transport
     Folly::folly
     ${GLOG_LIBRARIES}
@@ -157,7 +135,7 @@ endforeach()
 
 if (BUILD_SHARED_LIBS)
   # all but thrift since it's an interface
-  set_target_properties(thrift-core concurrency protocol transport async
+  set_target_properties(thrift-core concurrency transport async
     PROPERTIES VERSION ${PROJECT_VERSION}
     )
 endif()
@@ -166,7 +144,6 @@ install(
   TARGETS
     thrift-core
     concurrency
-    protocol
     transport
     async
     thrift
diff --git a/thrift/lib/cpp2/CMakeLists.txt b/thrift/lib/cpp2/CMakeLists.txt
index edc34cae5..e6a4673af 100644
--- a/thrift/lib/cpp2/CMakeLists.txt
+++ b/thrift/lib/cpp2/CMakeLists.txt
@@ -27,6 +27,7 @@ set(LIB_CPP2_HOME ${CMAKE_CURRENT_SOURCE_DIR})
 get_property(RpcMetadata-cpp2-SOURCES GLOBAL PROPERTY RpcMetadata-cpp2-SOURCES)
 get_property(metadata-cpp2-SOURCES GLOBAL PROPERTY metadata-cpp2-SOURCES)
 get_property(frozen-cpp2-SOURCES GLOBAL PROPERTY frozen-cpp2-SOURCES)
+get_property(reflection-cpp2-SOURCES GLOBAL PROPERTY reflection-cpp2-SOURCES)
 
 bypass_source_check("${metadata-cpp2-SOURCES}")
 add_library(
@@ -74,6 +75,7 @@ target_link_libraries(
     Folly::folly
 )
 
+bypass_source_check("${reflection-cpp2-SOURCES}")
 add_library(
   thriftprotocol
 
@@ -85,13 +87,24 @@ add_library(
   protocol/JSONProtocol.cpp
   protocol/Serializer.cpp
   protocol/VirtualProtocol.cpp
+
+  ../cpp/protocol/TDebugProtocol.cpp
+  ../cpp/protocol/TJSONProtocol.cpp
+  ../cpp/protocol/TBase64Utils.cpp
+  ../cpp/protocol/TProtocolException.cpp
+  ../cpp/protocol/TSimpleJSONProtocol.cpp
+  ${reflection-cpp2-SOURCES}
 )
+add_dependencies(thriftprotocol reflection-cpp2-target)
 target_link_libraries(
   thriftprotocol
   PUBLIC
+    fmt::fmt
     thrift
     Folly::folly
     wangle::wangle
+    ${GLOG_LIBRARIES}
+    thrift-core
 )
 
 add_library(
